language/ir-testsuite/tests/move/generics/phantom_params/struct_instantiation.mvir

```diff
@@ -0,0 +1,85 @@
+// This file contains tests checking that type arguments used in phantom position
+// (phantom arguments) are not considered when deriving the abilities for a struct.
+
+module M1 {
+    // This module checks that phantom arguments are not considered by checking against
+    // abilities required for specific bytecode operations.
+
+    struct NoAbilities { a: bool }
+    struct HasDrop<phantom T> has drop { a: bool }
+    struct HasCopy<phantom T> has copy { a : bool }
+    struct HasStore<phantom T> has store { a : bool }
+    struct HasKey<phantom T> has key { a : bool }
+
+    // `WriteRef` requires drop
+    f1(ref: &mut Self.HasDrop<Self.NoAbilities>) {
+        *move(ref) = HasDrop<Self.NoAbilities> { a: true };
+        return;
+    }
+
+    // `Pop` requires drop
+    f2() {
+        _ = HasDrop<Self.NoAbilities> { a: true };
+        return;
+    }
+
+    // Leaving value in local requires drop
+    f3(x: Self.HasDrop<Self.NoAbilities>) {
+        return;
+    }
+
+    // `CopyLoc` requires copy
+    f4(x: Self.HasCopy<Self.NoAbilities>): Self.HasCopy<Self.NoAbilities> * Self.HasCopy<Self.NoAbilities> {
+        return (copy(x), move(x));
+    }
+
+    // `MoveTo` requires key
+    f5(s: &signer, x: Self.HasKey<Self.NoAbilities>) {
+        move_to<HasKey<Self.NoAbilities>>(move(s), move(x));
+        return;
+    }
+
+    // `MoveFrom` requires key
+    f6(): Self.HasKey<Self.NoAbilities> acquires HasKey {
+        return move_from<HasKey<Self.NoAbilities>>(0x0);
+    }
+
+    // `Exists` requires key
+    f7(): bool {
+        return exists<HasKey<Self.NoAbilities>>(0x0);
+    }
+}
+
+//! new-transaction
+
+module M2 {
+    // This module checks that phantom arguments are not considered by checking against
+    // the abilities required for the fields of a struct.
+
+    import {{default}}.M1;
+
+    struct HasDrop has drop { a: M1.HasDrop<M1.NoAbilities> }
+    struct HasCopy has copy { a: M1.HasCopy<M1.NoAbilities> }
+    struct HasStore has store { a: M1.HasStore<M1.NoAbilities> }
+    struct HasKey has key { a: M1.HasStore<M1.NoAbilities> }
+}
+
+//! new-transaction
+
+module M3 {
+    // This module checks that phantom arguments are not considered by checking
+    // against ability constraints.
+
+    import {{default}}.M1;
+
+    struct S1<T1: drop, T2: copy, T3: store, T4: key> { a: bool }
+    struct S2 {
+        a: Self.S1<M1.HasDrop<M1.NoAbilities>, M1.HasCopy<M1.NoAbilities>, M1.HasStore<M1.NoAbilities>, M1.HasKey<M1.NoAbilities>>,
+    }
+
+    f1<T1: drop, T2: copy, T3: store, T4: key>() { return; }
+    f2() {
+        Self.f1<M1.HasDrop<M1.NoAbilities>, M1.HasCopy<M1.NoAbilities>, M1.HasStore<M1.NoAbilities>, M1.HasKey<M1.NoAbilities>>();
+        return;
+    }
```

Comment:
The constraint behavior here feels a bit different than the bytecode related checks above. Maybe split it out into a separate test?

Reply:
for me is a variation of the same. you are checking that a struct instantiation has the correct abilities by using them in different places that require them to have those abilities.

Reply:
Yeah but the actual code that you are testing is in different locations... You are right that they are conceptually the same, but tests are more helpful the smaller they are, since when something breaks it is easier to debug

Reply:
ok, makes sense. I put it into a separate file.